### ISSUE TITLE
execute reduxLoop.Cmd.action synchronously without notifying subscribers?

### ISSUE THREAD

**Issue by arnauorriols** [redux-loop/redux-loop#173]
I'm currently using `reduxLoop.Cmd.action` to trigger new actions right after the current action being processed in the reducer. Although it's fairly reasonable, I wasn't expecting that before any new action is processed by the reducer, the subscribers are notified (ie, the react views are rerendered). Right after the subscribers are notified, the new action is processed, and the subscribers notified again. 

My views are stateless, and this makes the intermediate notification to the subscribers redundant and costly from a performance point of view (in my real case, I have a chain of up to 3 chained actions). I'm currently debouncing these redundant notifications using `react-batched-subscribers` with `request-animation-frame`. However, this introduces some unnecessary complexity to my code, and I wonder if this could already be solved by redux-loop? I'm thinking of an option to `reduxLoop.Cmd.action` to process the actions synchronously, and only notify the subscribers until all the actions are processed. 

What are your thoughts about this?

**Comment by bdwain** [comment]
Hi @arnauorriols. I'm not sure we'd be able to prevent subscribers from being notified about an action that is dispatched, since that is part of the underlying redux functionality. But there are a few things that you can do to solve the issue.

1. Make all state updates in response to the original action
2. Combine your side effects into a single side effect (assuming you have some side effects).
3. Use shouldComponentUpdate to prevent updates to your react components.

My first thought when hearing how your reducer is structured is that you could cut down on the number of actions you have. Currently, it sounds like you have a chain of actions that directly dispatch other actions (using Cmd.action). But if no side effect or user interaction is happening to affect which actions are dispatched after the initial action, then it's all pre-determined and could be handled with a single dispatch/state update/rerender. 

In general I think it's helpful to think about actions as events that DID happen (button clicked) rather than describing an event that SHOULD happen (make request). Then your action history looks like an event log instead of being cluttered with a bunch of being filled with a bunch of intermediate actions that could be removed without affecting the state history. That's generally why I don't use Cmd.action very much unless I'm triggering an event in an unrelated piece of code (e.g. update state and then change the router location).

If you are using the intermediate actions to avoid repetition (e.g. BUTTON1_CLICKED and BUTTON2_CLICKED both trigger INTERMEDIATE_ACTION), one thing that has worked for me is to just handle multiple action types together.

```js
case BUTTON1_CLICKED:
case BUTTON2_CLICKED:
//return intermediateStuff(state, action);
```

From your description it sounds like you're not running any side effects (Cmd.run) as part of this chain of updates, but if I's misunderstanding, you could also chain your side effects into a single side effect function that calls all 3 side effects before returning. This is helpful if the different parts of a multi-step side effect don't change what is rendered until they are all done.

If all else fails, you can always use shouldComponentUpdate in your react component to prevent re-rendering until the final action is done being dispatched. But my thought is you should be able to avoid this if you use the first suggestion.

**Comment by arnauorriols** [comment]
> In general I think it's helpful to think about actions as events that DID happen (button clicked) rather > than describing an event that SHOULD happen (make request). Then your action history looks like an > event log instead of being cluttered with a bunch of being filled with a bunch of intermediate actions > that could be removed without affecting the state history.

This sounds interesting. I will have to think longer about your response, as it implies a different approach in respect of what the actions mean:

My reducer is structured so that each action defines *a possible mutation that can happen to the state of the view*. With that I mean that it's not structured mapping the user events, but mapping the state changes that can happen.

For example, considering a view that is a map, with some markers, and a panel where more information is shown when the user clicks the marker. When the user clicks the marker, we want to:
- Show in the panel the information about the marker
- Center the map

This is the state of the view (typescript);
```typescript
interface State {
    mapBounds: Bounds;
    selectedMarker: Mabey<Marker>;
}
```

And these would be the Actions:
```typescript
type Action
    = { type: "SELECT MARKER", marker: Marker }
    | { type: "CHANGE MAP BOUNDS", bounds: Bounds }
    ;
```

From the view, we dispatch:
```typescript
dispatch({type: "SELECT MARKER", marker: {...}});
```

And the reducer goes like this:
```typescript
switch(action.type) {
    case "SELECT MARKER":
        const newState = {
            ...state,
            selectedMarker: maybe.just(action.marker)
        };
        const newCmd = reduxLoop.Cmd.action({
            type: "CHANGE MAP BOUNDS",
            bounds: centerMap(action.marker)
        });
        return [newState, newCmd];

    case "CHANGE MAP BOUNDS":
        const newState = {
            ...state,
            mapBounds: action.bounds
        };
        return [newState, reduxLoop.Cmd.none];
}
```
The rationale is that it's up to the reducer (not the view) to know that when a marker is selected, the bounds of the map need to change *to maintain the state of the view consistent*. Also, under this design, each action describes, and therefore is responsible, for only a *single* mutation of the state. So, the action "SELECT MARKER" has to know that the state has to mutate further, but only as much as knowing which is the next mutation that has to happen. 

I'll think more about your response and answer back with my conclusions. Many thanks anyhow!

**Comment by bdwain** [comment]
Yup. So in your example, my suggestion would be to put all of that logic in the handler for SELECT MARKER. 

```js
switch(action.type) {
    case "SELECT MARKER":
        return {
            ...state,
            selectedMarker: maybe.just(action.marker),
            bounds: centerMap(action.marker) 
        };
     default:
        return state;;
}
```

now let's say you could change the map bounds another way (let's say when FOO is dispatched, where FOO is some arbitrary thing that can happen), and wanted to share the logic so that you didn't have to repeat it. You could do something like this

```js
function updateBounds(state, bounds){
  return {
     ...state,
     bounds
  };
}

switch(action.type) {
    case "SELECT MARKER":
        return updateBounds({
            ...state,
            selectedMarker: maybe.just(action.marker)
        }, centerMap(action.marker));

    case "FOO":
       return centerMap(state, action.bounds);

   case 
     default:
        return state;;
}
```

this scenario is pretty simple so not sharing the logic would probably be the easiest thing to do for it, but if the logic in updateBounds was not trivial, then that pattern could help.

**Comment by bdwain** [comment]
also just as a side note, i recommend using the [loop](https://github.com/redux-loop/redux-loop/blob/master/docs/ApiDocs.md#loopstate-cmd-any-cmd) function rather than constructing an array manually, because the 2 element array is just an implementation detail and not guaranteed to stay the same between updates.
